### ISSUE TITLE
main script integration of pipeline steps brainstorming 

### ISSUE THREAD

**Issue by F-said** [NDCLab/baseEEG#77]
## Main script steps

1. take in pipeline constants:
    - **One** user_params.json file exists for parameter passing for pipeline
        - [MNE Example ](https://github.com/NDCLab/mne-study-template/blob/master/config.py)
2. Run pipeline steps, passing constants as needed.
3. Write to annotations_preproc.log & output.log as pipeline steps execute
4. output files after pipeline completes: 
    - **Many** annotations_preproc.json files exists for each subject in BIDS file post-pipeline
    - **Many** output.log files for each subject in BIDS file post-pipeline

Please add on/remove as needed.

**Comment by yanbin-niu** [comment]
I was thinking, we can have a class for user_params. It has a whole bunch of parameters that we read from user_params.json file. Call `read_user_json` will read user_params.json file and return an instance to the main. Then in the main, we pass this instance to each feature (rather than passing the parameters of the instance to each feature). I agree @DMRoberts  's comments on the differences between mne and our script. It does not make sense in each our feature, we just use mne methods. So, I kind of like the idea that we pass the instance of our self-defined  `user_params` to each feature. Something like:

```
user_params = read_user_json()

raw = feature_io(user_params)
raw_icaed = preprocessing_ica(raw, user_params)
raw_bad_channel_removed = preprocessing_bad_channel(raw_icaed, user_params)
.......
```

is this [MNE Example](https://github.com/NDCLab/mne-study-template/blob/master/config.py) what we want? (reading user_params.json file and assigning values to constants) @georgebuzzell  Sorry if my mindset still stays on old JAVA days.

**Comment by georgebuzzell** [comment]
Thank you @F-said and @yanbin-niu ! 

I modified what @F-said posted above.  Note that changes were made to reflect the changing of "annotations" to "output", but also some other additions about the organization of things.

The preprocessing "pipeline" consists of:

1. A single user_params.json file that lists a series of constants that the user defines (or leaves set to the default values). This includes the path to the location of the data, filter settings, ICA settings, the location of where to write the data, etc.
2. A "Main script" that first reads in the user_params.json and creates a params dictionary containing all of the variable names and their values from user_params.json. The main script then calls a series of functions (features) that we have written. Each time the main script calls a feature function that we have written, it passes the params dictionary, the output dictionary, and the data. The main script will also directly call mne functions when there is no need for a separate feature function to write data or write to output_preproc.
3. A series of feature functions that we have written, that contain a mix of mne functions and custom code. Some are rather simple and more or less just call an mne function, some are more complicated, but they all follow the same standard format: each feature always receives the params dictionary and data, passed from the main script. The feature function then executes custom code and/or calls other mne functions using the relevant parameters pulled from the params dictionary passed. The feature function then calls a separate feature function for reading/updating the output_params.JSON file. the feature function finishes by returning the updated data to the main script. 
4. A Feature function dedicated to: 1) receiving parameters to be written (from one of our feature functions); 2) reading the most recent version of the output_params.JSON file for this participant; 3) updating and rewriting the output_params.JSON file with the parameters to be written.

The preprocessing pipeline produces:
- **Many** processed EEG files
- **Many** output_preproc.json files; one is created for each subject in BIDS file post-pipeline
- **Many** output.log files; one is created for each subject in BIDS file post-pipeline

**Comment by Jonhas** [comment]
@yanbin-niu @F-said Thank you for your insight! It made me think on the surface on how the pipeline should be structured. I don't think that a class would be necessary for user_params, since the class would not have any other functionality besides storing data nor would it be polymorphic or have encapsulated data. I don't think there is a need for abstraction either, since we are only going through some procedures within the pipeline. You can kinda resemble a struct (kinda of a non OOP class) in python through immutable tuples, but it is rather tedious and can be too complex with all of the data we may have in the json. 

Therefore, I believe its for the best to keep it as a dictionary, where we can then call the values needed such as user_params["key needed here"] throughout our pipeline. I recommend having the pipeline to be procedural rather than object-oriented. Procedural programming is based on the concept of the procedure call, where procedures (a type of routine or subroutine) simply contain a series of computational steps to be carried out, i.e, we break the code down into subroutines. In other words, it is a mechanism for structuring programs in a sequential manner, such as 
``` 
First I do x 
Second I do y
Thirdly I do Z
```
It is often simpler than OOP, (I bring this up since Java is almost a pure object oriented programming language) and offers the benefits of modularity. 

Essentially, since there will be one main script, we want to use references of other modules to perform our desired action. We would need functions to write into output.log files, write to output_preproc.json files, etc, and these can all be expressed through a module. The procedures within the model will simply then be used on the main script. Since right now all of the functionality is written as independent scripts to assure we satisfied our requirements, we would need to change these scripts to be modules. As a result, we would then be able to properly communicate with the main script. 

@georgebuzzell I agree with everything you've mentioned. No comments here.

**Comment by yanbin-niu** [comment]
@Jonhas Thanks for the insights! Agree! If our user_params just simply need a place to store the data, a dictionary is definitely easy-to-use and convenient.

Also, I agree with the procedural programming, which is what said by @georgebuzzell ! Finally, @georgebuzzell everything sounds good to me, too.

**Comment by DMRoberts** [comment]
One suggestion for passing the parameters in the JSON file, building on what @yanbin-niu mentioned, is to use argument unpacking (**) to pass all of the defined parameters to each function. That way, the functions themselves don't need knowledge of the the JSON / dict format, they just receive each parameter by name, and could also be used outside of the pipeline, for example:

```
def function_one(param_one, param_two):
    return param_one + param_two

def function_two(x, y):
    return x/y

user_params = {'function_one': {'param_one': 17, 'param_two': 2},
               'function_two': {'y': 2, 'x': 4} }

# give all parameters listed under 'function_one' or 'function_two' to the appropriate function
result_one = function_one(**user_params['function_one'])
result_two = function_two(**user_params['function_two'])

# other parameters can also be passed by name (for example if we wanted to pass the MNE EEG object as first param)
# function w/ three parameters
def function_three(x, y, z):
    return x/y + z

# specify the first parameter outside of the dictionary
result_three = function_three(z=2, **user_params['function_two'])
```

There may also be a way to call the functions needed from the string representations within the JSON / dictionary, so the main script would essentially just be iterating over the functions / parameters as indicated in the JSON / dictionary.

**Comment by Jonhas** [comment]
@DMRoberts this is a pretty intuitive way to look at it. When passing ** it allows us to unpack dictionaries, giving us the ability to pass an unspecified number of arguments to the function. This is similar to passing a pointer to a pointer of a list and then iterating through the single pointer. I agree that this is a good way on handling passing around json data.

**Comment by georgebuzzell** [comment]
Thank you, @F-said @yanbin-niu @Jonhas and @DMRoberts !

@Jonhas I very much agree that we should use a procedural, as opposed to an object-oriented approach. Thank you for laying out the description of procedural; very helpful and much appreciated! One clarification that would be helpful for me. Can you please explain "modules" just a bit more?

@yanbin-niu thank you for your comments, questions, suggestions both on here and on Slack. They helped me to refine how I think about this as well!

@DMRoberts I really like your suggestion of how the functions will interact with the json/dict in a way that does not require the function to have knowledge of the Json/dict format. The reason this is really great is that this lines up with suggestions by @yanbin-niu and @Jonhas to set things up such that, in the future when we invariably want to update/change/upgrade particular functions, we can do so without worrying about changing the main script, or the other functions. I very strongly agree with this general approach suggested by all of you, and it sounds like the specific implementation that you have suggested, @DMRoberts would be a great way to do this.  

OK, from what I can tell, it seems like we have consensus, at least amongst me @Jonhas @yanbin-niu and @DMRoberts on a plan for how to move forward. @F-said can you please review these comments here and comment on whether you agree with everything or have additional suggestions for how to improve?

@F-said if you generally agree with all of this here, can you please do the following: 1) say that here (or suggest any refinements); 2) branch off of dev and modify the current roadmap to include a description of the plan outlined here; 3) do a PR and add me @yanbin-niu @Jonhas and @DMRoberts all as assigned reviewers?

**Comment by F-said** [comment]
@georgebuzzell Yep, everything looks great! Definitely learning more about Python's flexibility as well 😅

Taking in what @DMRoberts @Jonhas @yanbin-niu commented, I see that two ways of fulfilling the user parameters were discussed: 

1. Abstract away the entire pipeline to the `user_params` JSON file:

    * As @DMRoberts mentioned, we could set up the JSON file to list pipeline steps as string function names and associate them with their respective parameters: 
    ex:
    ```json
    {
        "filter": {
            "highPass": [0.3],
            "lowpass": [50]
        }, 
        "segment": {
            "tmin": [-0.2],
            "tmax": [0.5]
        }
    }
    ```
    
    * This would completely abstract away the programming of the pipeline. A user that would want to modify pipeline steps could simply work with the JSON file.
    * The main script would iterate over this file and use the `getattr()` method to call the functions from strings. 
    ```
    for function_name in user_params:
        getattr(moduleName, function_name)(**user_params[function_name])
    ```
    
2. Create explicit pipeline function calls in the main script:

    * The responsibility of the `user_params.json` file would be returned to only providing parameters.
    
    * While still keeping the main script procedural, perhaps we could delegate the loading of `user-params.json` dictionary into a separate script that's imported with the other pipeline steps. That way the main script is responsible for *only* executing the pipeline. 
    ```python
    from load_data import user_params 
    
    import filter
    import segment
    import bad-channels
    
    def main():
        filter.filter(**user_params["filter"])
        # and so on
   ```
   
Lastly, it looks like the README of baseEEG is quickly scaling up as development progresses. Would it be acceptable to create a new `contributing.md` file that contains all the details project contributors need while relegating `readme.md` to be a succinct description of the project?

The sections of each file would follow:

### Contributing.md

* Git workflow
* Issues
* CI Tests
* Input/Output file structure 
* Example Files
* Reminders

### readme.md

* Project description
* Roadmap
* Input/Output file usage and purpose

**Comment by Jonhas** [comment]
@georgebuzzell Modules can be seen as a code library. It is a file containing a set of functions you want to include in your application. A module can define functions, classes and variables. In our case, we would only want to define functions. Every time we are importing mne or mne_bids, we are importing the _mne module_ to the application. Essentially, a module allows us to logically organize your python code, and is loaded only once, regardless of the number of times it is imported. Therefore, we would have something similar:
```python3
def func1():
    # stuff here 

def func2():
    # stuff here 

global_variable = 3 
another_global_variable = [5,3,6,9]

```
so on and so forth. The benefits of modules:

1. Simplicity, since it focuses on a small portion of the problem
2. Re-usability 
3. Scoping. Modules establish a separate namespace (which is a declarative region that provides a scope to the identifiers (the names of types, functions, variables, etc) inside it), which is defined by the module and helps to avoid collisions between other identifiers that may share the same name.  

@F-said I agree that an additional Contributing.md may be necessary to make the readme.md more readable.

**Comment by georgebuzzell** [comment]
@F-said this all sounds great to me!

**Comment by georgebuzzell** [comment]
Thank you, @Jonhas !!

**Comment by georgebuzzell** [comment]
Closing this issue, as discussion is complete and a plan has been decided (see above and recording of call on zoom). Have created a new issue to update the readme with the final plan: https://github.com/NDCLab/baseEEG/issues/87
